### PR TITLE
[RFC] Introduce SobolEngine

### PR DESCRIPTION
**Pull request by vishwakftw**
1. Added the native implementations to ATen
2. Created a basic interface for SobolEngine for Python

Todo:
- Compile (done)
- Tests (unscrambled, done)
- Fix scrambling (done)
- Tests (scrambled, done)
- Docs (done)

cc: @Balandat

### CODE REVIEW

File: torch/sobolengine.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+        self.seed = seed
+        self.scramble = scramble
+        self.dimension = dimension
+
+        self.sobolstate = torch.zeros(dimension, self.MAXBIT, dtype=torch.int)
+        torch._sobol_engine_initialize_state(self.sobolstate, self.dimension)
+
+        if scramble:
+            self.shift = torch.mv(torch.randint(2, (self.dimension, self.MAXBIT)),
```
Comment (vishwakftw) [inline torch/sobolengine.py]: There are two ways of seeding: using a `torch.Generator` object and then seeding to, so that the seeding is "local". Or you could modify the global seed as well.

Clearly, option 1 is the better approach, but I am not sure about it. There is no documentation of how to obtain `torch.Generator` objects to pass to random tensor factory functions. Once I work out the details, I will add it.

Comment (vishwakftw) [reply torch/sobolengine.py]: This has been taken care of. Scrambling should work now.

File: aten/src/ATen/native/SobolEngineOps.cpp [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+  // For every tensor in the list of tensors, the diagonals are made 1
+  // Require the rows of each of the matrices in `ltm`.
+  // Why is this caching these rows separately okay?
+  // A simple calculation show the number of slices to be dimension * MAXBIT * MAXBIT
+  // Instead, by performing `dimension` chunking operations (1 per lower triangular
+  // matrix), we can save some time, since MAXBIT = 30.
+  Tensor eye_maxbit = at::native::eye(MAXBIT, at::ScalarType::Byte);
+  std::vector<std::vector<Tensor>> ltm_rows;
+  for (int64_t d = 0; d < dimension; ++d) {
+    ltm_rows.emplace_back(at::native::chunk(at::native::where(eye_maxbit,
```
Comment (vishwakftw) [inline aten/src/ATen/native/SobolEngineOps.cpp]: `at::where` is complaining that the first argument should be of type `Variable` but it is `at::kByte` instead. I am not sure how to fix this.

**Comment by vishwakftw** [comment]
@Balandat: an update - the unscrambled `SobolEngine` is ready

File: aten/src/ATen/native/SobolEngineOps.cpp [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+    num_generated++;
+  }
+  return wquasi;
+}
+
+/// This is an implicit function used for randomizing the state variables of the.
+/// `SobolEngine`. Arguments are a randomized `sobolstate` state variables
+/// and a list of random lower triangular matrices consisting of 0s and 1s. `dimension` is
+/// passed explicitly again.
+Tensor _sobol_engine_scramble(const Tensor& sobolstate, TensorList ltm, int64_t dimension) {
```
Comment (vishwakftw) [inline aten/src/ATen/native/SobolEngineOps.cpp]: There seems to be something wrong in this implementation that I am not able to figure out. Some help will be appreciated :).

**Comment by vishwakftw** [comment]
Scrambling is very slow, and I am currently looking at this issue. However, the whole PR is functional, meaning checking out this branch and building torch will give a working implementation.

**Comment by vishwakftw** [comment]
Scrambling is pretty fast now, courtesy of a Hadamard (element-wise) product to compute `cdot_pow2` faster.
Closing in favour of https://github.com/pytorch/pytorch/pull/10505 .
